### PR TITLE
Prepare to support parquet row group skipping

### PR DESCRIPTION
**Pull request by scovich**
In preparation for https://github.com/delta-incubator/delta-kernel-rs/pull/362 that actually implements parquet row group skipping, here we make various preparatory changes that can stand on their own:
* Plumb the predicates through to the parquet readers, so that they can easily start using them
* Add and use a new `Expression::is_not_null` helper that does what it says
* Factor out `replay_for_XXX` methods, so that log replay involving push-down predicates can be tested independently. 
* Don't involve <n>.json in log replay if <n>.checkpoint.parquet is available

This should make both changes easier to review.

### CODE REVIEW

File: ffi/src/engine_funcs.rs [hunk 1]
```diff
@@ -122,6 +122,7 @@ fn read_parquet_file_impl(
         last_modified: file.last_modified,
         size: file.size,
     };
+    // TODO: Plumb the predicate through the FFI?
```
Comment (zachschuermann) [inline ffi/src/engine_funcs.rs:125]: created https://github.com/delta-incubator/delta-kernel-rs/issues/382

**Review by zachschuermann — APPROVED** [review]
nice thanks ryan, really like the new `replay_for_*` LGTM!

File: kernel/tests/read.rs [hunk 2]
```diff
@@ -744,7 +748,7 @@ fn predicate_on_number_with_not_null() -> Result<(), Box<dyn std::error::Error>>
         "./tests/data/basic_partitioned",
         Some(&["a_float", "number"]),
         Some(Expression::and(
-            Expression::not(Expression::column("number").is_null()),
+            Expression::column("number").is_not_null(),
```
Comment (nicklan) [inline kernel/tests/read.rs:751]: so much nicer :)

**Review by nicklan — APPROVED** [review]
lgtm! thanks

**Pull request merged**